server/market/order/order.go

```diff
@@ -0,0 +1,326 @@
+// Copyright (c) 2019, The Decred developers
+// See LICENSE for details.
+
+// Package order defines the Order and Match types used throughout the DEX.
+package order
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"math"
+	"time"
+
+	"github.com/decred/dcrd/crypto/blake256"
+	"github.com/decred/dcrdex/server/account"
+)
+
+// OrderIDSize defines the length in bytes of an OrderID.
+const OrderIDSize = blake256.Size
+
+// OrderID is the unique identifier for each order.
+type OrderID [OrderIDSize]byte
+
+// String returns a hexadecimal representation of the OrderID. String implements
+// fmt.Stringer.
+func (oid OrderID) String() string {
+	return hex.EncodeToString(oid[:])
+}
+
+// OrderType distinguishes the different kinds of orders (e.g. limit, market,
+// cancel).
+type OrderType uint8
+
+// The different OrderType values.
+const (
+	LimitOrderType OrderType = iota
+	MarketOrderType
+	CancelOrderType
+)
+
+// TimeInForce indicates how limit order execution is to be handled. That is,
+// when the order is not immediately matched during processing of the order's
+// epoch, the order may become a standing order or be revoked without a fill.
+type TimeInForce uint8
+
+// The TimeInForce is either ImmediateTiF, which prevents the order from
+// becoming a standing order if there is no match during epoch processing, or
+// StandingTiF, which allows limit orders to enter the order book if not
+// immediately matched during epoch processing.
+const (
+	ImmediateTiF TimeInForce = iota
+	StandingTiF
+)
+
+// Order specifies the methods required for a type to function as a DEX order.
+// See the concrete implementations of MarketOrder, LimitOrder, and CancelOrder.
+type Order interface {
+	// ID computes the Order's ID from its serialization as per the spec.
+	ID() OrderID
+
+	// UID gives the string representation of the order ID. It is named to
+	// reflect the intent of providing a unique identifier.
+	UID() string
+
+	// Serialize marshals the order as per the spec.
+	Serialize() []byte
+
+	// SerializeSize gives the length of the serialized order in bytes.
+	SerializeSize() int
+
+	// Type indicates the Order's type (e.g. LimitOrder, MarketOrder, etc.).
+	Type() OrderType
+
+	// Time returns the Order's server time, when it was received by the server.
+	Time() int64
+
+	// Remaining computes the unfilled amount of the order.
+	Remaining() uint64
+}
+
+// An order's ID is computed as the Blake-256 hash of the serialized order.
+func calcOrderID(order Order) OrderID {
+	return blake256.Sum256(order.Serialize())
+}
+
+// UTXO is the interface required to be satisfied by any asset's implementation
+// of a UTXO type.
+type UTXO interface {
+	TxHash() []byte
+	Vout() uint32
+	Serialize() []byte
+	SerializeSize() int
+}
+
+// Prefix is the order prefix containing data fields common to all orders.
+type Prefix struct {
+	AccountID  account.AccountID
+	BaseAsset  uint32 // must be 0 for CancelOrder
+	QuoteAsset uint32 // must be 0 for CancelOrder
```

Comment:
I imagine the DEX will be in charge of mapping the integer identifier to the ticker symbol string.  Perhaps the "asset object" returned to the client in response to the `assets` RPC in https://github.com/decred/dcrdex/blob/master/spec/README.mediawiki#exchange-variables can be updated with the integer ID.

My thinking about this `Prefix` struct is that it's primary purpose is byte serialization for computing order ID and storing the order data compactly, while the ticker symbol strings would come in with JSON serialization, and that would be something handled by something higher up like the dex or communications hub.

I suppose I'd be open to having these fields as arrays (i.e. `[5]byte`), but I wanted to avoid a string here.

Reply:
`[5]byte` is interesting. We need it to be repeatable for future mesh interoperability. If we wanted an int, we could potentially use the BIP-0044 coin type index.
 
https://github.com/satoshilabs/slips/blob/master/slip-0044.md

Reply:
BIP-0044 coin types makes a lot of sense.  I'd go with that as a convention unless there are strong architectural benefits of putting the ticker characters in the order.Prefix object, but I think the ticker mapping can exist at a higher level in the dex.

Reply:
See https://github.com/decred/dcrdex/pull/29

Reply:
An itch that I still can't quite scratch is that these two asset IDs don't apply to a cancel order, or don't have to at least.
- Should cancel order prefixes be validated to ensure these values are as expected for the order in question (not much point of that)?
- Should they be required to both be some other sentry value like uint32(-1)?
- Should the asset types be removed from the prefix to the MarketOrder (and thus LimitOrder too)?

Reply:
I can see it either way. I like to think that if I was handed a serialized cancel order, I would be able to tell what market it's for, but you're right that it's not a necessity.

Reply:
For now, we'll set these asset types for cancel orders, but I'm not sure if we'll want to bother validating them for incoming cancel orders.